### ISSUE TITLE
Consider introducing holes and undefined

### ISSUE THREAD

**Issue by razcore-art** [purescript-contrib/purescript-book#233]
I remember when I first started working with Purescript and getting more serious about it that one of the most frustrating parts was that I couldn't get to test or get the compiler to help me figure things out as I first had to reason the entire implementation of the function.

I think the idea of [typed holes](https://github.com/purescript/documentation/blob/6b9cdddb687dd8b4b580ca44a94a04680bda9149/errors/HoleInferredType.md) and [undefined](https://pursuit.purescript.org/packages/purescript-undefined/1.0.2/docs/Undefined#v:undefined) should be introduced fairly early with examples. Using these features greatly increases both development and lowers the stress of not figuring out functions "from the first try".

**Comment by milesfrain** [comment]
Maybe we could add an "extra credit" or "recommended reading" section to the end of some chapters (along with some related exercises).

I don't want to duplicate any [typed holes or type-directed search](https://github.com/paf31/24-days-of-purescript-2016/blob/master/23.markdown) documentation, but I could envision some exercises to accompany a link to that article.

This is my first exposure to `undefined` in PS. Can't think of any exercises for that at the moment.

What chapter do you think would be a good time to learn about these concepts?

**Comment by razcore-art** [comment]
A recommended reading would be a great way to introduce some of these concepts as well.

`undefined` in PS is basically the same thing as `undefined` in Haskell, it's the "bottom" element.

See [docs](https://github.com/purescript/documentation/blob/6b9cdddb687dd8b4b580ca44a94a04680bda9149/language/Differences-from-Haskell.md#error-and-undefined) and the link to the issue included there for an example usage.

It's not so much that it should have an accompanying exercise, but rather it should be explained how it's useful to supress complier errors. Perhaps `undefined` isn't as useful as typed holes, cause at the end of the day, as they discuss in the linked issue, `undefined` can be dangerous if you don't use it sparingly.

---

I'd see typed holes introduced either in the functions chapter (Ch3) or at the very least before the type class chapter (Ch6). Because Ch6 is where stuff gets exponentially difficult for a beginner to understand really new concepts that could get in the way when solving the exercises. And typed holes could be very handy with the more complex exercises.

**Comment by hdgarrood** [comment]
We actually deliberately omitted `undefined` from the core libraries, because it's a liability. I don't think we should encourage its use in the book for the same reason.

**Comment by razcore-art** [comment]
Yeah, I agree with that after some more thought. But typed holes should definitely be covered somewhere
